packages/core/contracts/oracle/implementation/DesignatedVotingV2.sol

```diff
@@ -0,0 +1,120 @@
+        _setWithdrawRole(uint256(Roles.Owner));
+
+        finder = FinderInterface(finderAddress);
+    }
+
+    /****************************************
+     *   VOTING AND REWARD FUNCTIONALITY    *
+     ****************************************/
+
+    /**
+     * @notice Forwards a commit to Voting.
+     * @param identifier uniquely identifies the feed for this vote. EG BTC/USD price pair.
+     * @param time specifies the unix timestamp of the price being voted on.
+     * @param hash the keccak256 hash of the price you want to vote for and a random integer salt value.
+     */
+    function commitVote(
+        bytes32 identifier,
+        uint256 time,
+        bytes memory ancillaryData,
+        bytes32 hash
+    ) external onlyRoleHolder(uint256(Roles.Voter)) {
+        _getVotingContract().commitVote(identifier, time, ancillaryData, hash);
+    }
+
+    /**
+     * @notice Forwards a batch commit to Voting.
+     * @param commits struct to encapsulate an `identifier`, `time`, `hash` and optional `encryptedVote`.
+     */
+    function batchCommit(VotingAncillaryInterface.CommitmentAncillary[] calldata commits)
+        external
+        onlyRoleHolder(uint256(Roles.Voter))
+    {
+        _getVotingContract().batchCommit(commits);
+    }
+
+    /**
+     * @notice Forwards a reveal to Voting.
+     * @param identifier voted on in the commit phase. EG BTC/USD price pair.
+     * @param time specifies the unix timestamp of the price being voted on.
+     * @param price used along with the `salt` to produce the `hash` during the commit phase.
+     * @param salt used along with the `price` to produce the `hash` during the commit phase.
+     */
+    function revealVote(
+        bytes32 identifier,
+        uint256 time,
+        int256 price,
+        bytes memory ancillaryData,
+        int256 salt
+    ) external onlyRoleHolder(uint256(Roles.Voter)) {
+        _getVotingContract().revealVote(identifier, time, price, ancillaryData, salt);
+    }
+
+    /**
+     * @notice Forwards a batch reveal to Voting.
+     * @param reveals is an array of the Reveal struct which contains an identifier, time, price and salt.
+     */
+    function batchReveal(VotingAncillaryInterface.RevealAncillary[] calldata reveals)
+        external
+        onlyRoleHolder(uint256(Roles.Voter))
+    {
+        _getVotingContract().batchReveal(reveals);
+    }
+
+    /**
+     * @notice Forwards a reward retrieval to Voting.
+     * @dev Rewards are added to the tokens already held by this contract.
+     * @return amount of rewards that the user should receive.
+     */
+    function retrieveRewards() public onlyRoleHolder(uint256(Roles.Voter)) returns (uint256) {
+        StakerInterface voting = StakerInterface(address(_getVotingContract()));
+        uint256 rewardsMinted = voting.withdrawRewards();
+        IERC20(address(voting.votingToken())).approve(address(voting), rewardsMinted);
+        voting.stake(rewardsMinted);
+    }
```

Comment:
Is there any way for a user to pull the rewards into the contract rather than restaking them, so they don't have to always deal with a lockup?

IIRC, this was the reason we wanted the rewards to be disbursed to the user by default rather than being restaked by default.

Reply:
I chose to do it like this as this now behaves the _same_ as the current Designated voting contract. claiming rewards in the current design just sends them to the designated voting contract. in this implementation by `retrieveRewards` re-staking, this behaves the same as the current design.

if we wanted voters to be able to access these this has implications for how we use this as a vesting mechanism and I don't think this decision has been made yet.